File: 2560-House Robber IV.py
class Solution:
    def minCapability(self, nums: List[int], k: int) -> int:
        left, right = 1, max(nums)
        
        def count(mx):
            tot = 0
            picked = False
            for i in range(len(nums)):
                if nums[i] <= mx and not picked:
                    picked = True
                    tot += 1
                elif picked:
                    picked = False
            return tot >= k
            
        
        while left + 1 < right:
            mid = left + (right-left)//2
            if count(mid):
                right = mid
            else:
                left = mid
        
        if count(left):
            return left
        else:
            return right

File: C++/2560-House Robber IV.cpp
class Solution {
public:
    int k, N;
    bool isOK(vector<int>& nums, int maxVal) {
        int cnt = 0;

        for (int i=0; i<N; i++) {
            if (nums[i] <= maxVal) {
                if (++cnt >= k) return true;
                i++;
            }
        }

        return false;
    }
    int minCapability(vector<int>& nums, int k) {
        this->k = k;
        this->N = nums.size();
        int mn = INT_MAX, mx = 0;
        for (int i=0; i<N; i++) {
            mn = min(mn, nums[i]);
            mx = max(mx, nums[i]);
        }

        int left = mn, right = mx;
        while (left + 1 < right) {
            int mid = left + (right-left)/2;
            if (isOK(nums, mid)) {
                right = mid;
            } else {
                left = mid;
            }
        }

        if (isOK(nums, left)) {
            return left;
        }
        return right;
    }
};